model/models.py

```diff
@@ -186,17 +235,351 @@ def __call__(self, inputs):
+    dtype: the dtype of the computation (default: float32)
+    precision: numerical precision of the computation see `jax.lax.Precision`
+      for details.
+
+  Returns:
+    Output of shape `[batch..., q_length, num_heads, v_depth_per_head]`.
+  """
+  assert key.ndim == query.ndim == value.ndim, 'q, k, v must have same rank.'
+  assert query.shape[:-3] == key.shape[:-3] == value.shape[:-3], (
+      'q, k, v batch dims must match.')
+  assert query.shape[-2] == key.shape[-2] == value.shape[-2], (
+      'q, k, v num_heads must match.')
+  assert key.shape[-3] == value.shape[-3], 'k, v lengths must match.'
+
+  # compute attention weights
+  attn_weights = dot_product_relative_attention_weights(
+    query, key, bias, relative_ids, relative_embeddings,
+    relative_biases, broadcast_dropout, dropout_rng, dropout_rate,
+    deterministic, dtype, precision)
+
+  # return weighted sum over values for each query position
+  return jnp.einsum('...hqk,...khd->...qhd', attn_weights, value,
+                    precision=precision)
+
+
+class MultiHeadDotProductRelativeAttention(Module):
+  """Multi-head dot-product attention.
+
+    Attributes:
+      num_heads: number of attention heads. Features (i.e. inputs_q.shape[-1])
+        should be divisible by the number of heads.
+      dtype: the dtype of the computation (default: float32)
+      qkv_features: dimension of the key, query, and value.
+      out_features: dimension of the last projection
+      broadcast_dropout: bool: use a broadcasted dropout along batch dims.
+      dropout_rate: dropout rate
+      deterministic: if false, the attention weight is masked randomly
+        using dropout, whereas if true, the attention weights
+        are deterministic.
+      precision: numerical precision of the computation see `jax.lax.Precision`
+        for details.
+      kernel_init: initializer for the kernel of the Dense layers.
+      bias_init: initializer for the bias of the Dense layers.
+      use_bias: bool: whether pointwise QKVO dense transforms use bias.
+      attention_fn: dot_product_attention or compatible function. Accepts
+        query, key, value, and returns output of shape
+        `[bs, dim1, dim2, ..., dimN,, num_heads, value_channels]``
+      decode: whether to prepare and use an autoregressive cache.
+  """
+  num_heads: int
+  dtype: Dtype = jnp.float32
+  qkv_features: Optional[int] = None
+  out_features: Optional[int] = None
+  broadcast_dropout: bool = True
+  dropout_rate: float = 0.
+  deterministic: Optional[bool] = None
+  precision: Any = None
+  kernel_init: Callable[[PRNGKey, Shape, Dtype], Array] = default_kernel_init
+  bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = zeros
+  use_bias: bool = True
+  attention_fn: Callable[[Array, Array, Array], Array] = dot_product_relative_attention
+  decode: bool = False
+  relative_radius: Optional[int] = None
+  relative_bias: bool = False
+  
+  def setup(self):
```

Comment:
Hmm, I don't know if it makes a difference. But when I implemented relative attention, I got JAX errors if I initialized the "nn.Embed" layers in the "setup" method. I just create them in the __call__ method, like the other layers.